### ISSUE TITLE
cKDTree reduce redundant distance code.

### ISSUE THREAD

**Issue by rainwoodman** [scipy/scipy#6941]
Currently, query_knn and the rest of cKDTree uses two separate sets of code for calculating distance.

They can be merged without much performance penalty. This may help clearing up https://github.com/scipy/scipy/issues/5657 and https://github.com/scipy/scipy/pull/6663

**Comment by sturlamolden** [comment]
There is also redundancy with respect to the C code behind ```scipy.spatial.distance```.

**Comment by sturlamolden** [comment]
I agree that this should be cleaned up. We don't need three (or more?) copies of the same code in ```scipy.spatial.*```.

**Comment by rainwoodman** [comment]
This is being worked out on #6945 . I think I gradually realized there wasn't much redundancy in ckdtree itself. We could probably wrap spatial.distance to the same code as a separate PR?

**Comment by StanczakDominik** [comment]
This could probably be closed now, given the merging of #6945 back in '17 :)
